src/ServiceControl.Hosting/WindowsServiceWithRequestTimeout.cs

```diff
@@ -0,0 +1,39 @@
+namespace ServiceControl.Hosting;
+
+using System;
+using System.Runtime.Versioning;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Hosting.WindowsServices;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+[SupportedOSPlatform("windows")]
+sealed class WindowsServiceWithRequestTimeout : WindowsServiceLifetime
+{
+    static readonly TimeSpan CancellationDuration = TimeSpan.FromSeconds(5);
+    readonly HostOptions hostOptions;
+
+    // TODO: This constructor should not be needed and exist for backwards compability in the runtime
+    // public WindowsServiceWithRequestTimeout(IHostEnvironment environment, IHostApplicationLifetime applicationLifetime, ILoggerFactory loggerFactory, IOptions<HostOptions> optionsAccessor)
+    //     : this(environment, applicationLifetime, loggerFactory, optionsAccessor, Options.Create(new WindowsServiceLifetimeOptions()))
+    // {
+    // }
+
+    public WindowsServiceWithRequestTimeout(IHostEnvironment environment, IHostApplicationLifetime applicationLifetime, ILoggerFactory loggerFactory, IOptions<HostOptions> optionsAccessor, IOptions<WindowsServiceLifetimeOptions> windowsServiceOptionsAccessor)
+        : base(environment, applicationLifetime, loggerFactory, optionsAccessor, windowsServiceOptionsAccessor)
+    {
+        hostOptions = optionsAccessor.Value;
+    }
+
+    protected override void OnStop()
+    {
+        RequestAdditionalTime(hostOptions.ShutdownTimeout + CancellationDuration);
+        base.OnStop();
+    }
+
+    protected override void OnShutdown()
+    {
+        RequestAdditionalTime(hostOptions.ShutdownTimeout + CancellationDuration);
+        base.OnShutdown();
+    }
```

Comment:
@ramonsmits, according to the [documentation](https://learn.microsoft.com/en-us/dotnet/api/system.serviceprocess.servicebase.requestadditionaltime?view=net-9.0-pp#system-serviceprocess-servicebase-requestadditionaltime(system-timespan)), OnShutdown is ignored.

```suggestion
```